specs/sds_detection_modules.adoc

```diff
@@ -0,0 +1,301 @@
+    sds_pkg_name = fields.StrField("sds_pkg_name")
+    sds_pkg_version = fields.StrField("sds_pkg_version")
+    roles = fields.StrField("roles")
+    cluster_attrs = fields.DictField("cluster_attrs", {'str': 'str'})
+
+    def render(self):
+        self.__name__ = self.__name__ % self.node_id
+        return super(NodeContext, self).render()
+```
+
+* The definitions file would get additional details for detected SDS details as
+below for Node_context
+
+```
+data = """---
+namespace.tendrl.node_agent:
+  objects:
+    Cpu:
+    ........
+    Node_context:
+      attrs:
+        machine_id:
+          help: "Unique /etc/machine-id"
+          type: String
+        fqdn:
+          help: "FQDN of the Tendrl managed node"
+          type: String
+        node_id:
+          help: "Tendrl ID for the managed node"
+          type: String
+        cluster_id:
+          help: Id of the cluster to which node belongs to
+          type: String
+        sds_pkg_name:
+          help: Storage system package name
+          type: String
+        sds_pkg_version:
+          help: Storage system package version
+          type: String
+        roles:
+          help: Roles of the node
+          type: String
+        cluster_attrs:
+          help: Additional cluster specific attributes
+          type: json
+      enabled: true
+      value: nodes/$Node_context.node_id/Node_context
+    ...............
+```
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+None
+
+==== Notifications/Monitoring impact:
+
+None
+
+==== Tendrl/common impact:
+
+* Common module should define a generic class which takes care of mounting a
+plugin would be something like
+
+```
+class PluginMount(type):
+    def __init__(cls, name, bases, attrs):
+        if not hasattr(cls, 'plugins'):
+            cls.plugins = []
+        else:
+            cls.register_plugin(cls)
+
+    def register_plugin(cls, plugin):
+        instance = plugin()
+        cls.plugins.append(instance)
+```
+
+==== Tendrl/node_agent impact:
+
+* The node agent component need to define interfaces which extend from commons
+AutoDetectPlugin and add specific functions as below
+
+```
+class SDSDetectPlugin(AutoDetectPlugin):
+    @abstractmethod
+    def detect_storage_system(self):
+        raise NotImplementedError()
+```
+
+* The node agent component would introduce two python modules for ceph and
+gluster. These would be specific classes which extend from above SDSDetectPlugin
+and provide the actual implementation for the function
+
+```
+class DetectCephStorageSystem(SDSDetectPlugin):
+    def detect_storage_system(self):
+        # Implement the logic to get the details of underlying storage system
+        # 1. Figure out if ceph bits installed and set the type accordingly
+        # 2. Figure the version of ceph bits installed and accordingly set the
+        # version
+        # 3. Figure out the role of the storage node (by figuring out if MON or
```

Comment:
Wrong. One detection module detects only one specific role. Which means that there would be a separate module for ceph mon and ceph osd. Only the mon module would query for the cluster map to get the peer list, osd module would not. Each of these modules assigns it's own specific tag if the detection is successful.